packages/apollo-codegen-core/src/compiler/index.ts

```diff
@@ -222,6 +244,8 @@ class Compiler {
     this.options = options;
 
     this.typesUsedSet = new Set();
+    this.unionTypesSet = new Set();
+    this.interfaceTypesMap = new Map();
```

Comment:
why are the union types a set and the interface types a map?

Reply:
@designatednerd There's a slight difference in the details that union types and interface types give us access to. For the union types, we want to list the union type name, and the types that make up that union. `graphql-js` supports this out of the box by giving `GraphQLUnionType`'s a `types` property, which holds the array of types that make up the union. Because of this, we can just store the `GraphQLUnionType`'s in a `Set`. When the `serializeToJSON` function is then called to write out the IR as JSON, we can access the union types we want to show by calling `getTypes()` on each `GraphQLUnionType` itself.

`GraphQLInterfaceType`'s are handled a bit differently though. `graphql-js` doesn't provide a `types` property (or something similar) that gives us easy access to every type using a specific interface. We have to get this ourselves. There was already a helper method setup in the `Compiler` class (in `index.ts`) to do this, called `possibleTypesForType`. Since it's already there in the compiler, I think it makes sense to leverage it, and pass both the interface type, and the types using that interface, through to the `serializeToJSON` function using a `Map`.

I'm happy to handle this differently of course. Thanks!

Reply:
was mostly curious, thanks for the explanation